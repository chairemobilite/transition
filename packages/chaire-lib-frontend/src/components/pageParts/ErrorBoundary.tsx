/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import Collapsible from 'react-collapsible';
import { withTranslation, WithTranslation } from 'react-i18next';

export type ErrorProps = WithTranslation & {
    key: string;
};

type ErrorState = {
    hasError: boolean;
    error?: Error;
    info?: React.ErrorInfo;
};

export class ErrorBoundary extends React.Component<React.PropsWithChildren<ErrorProps>, ErrorState> {
    constructor(props: React.PropsWithChildren<ErrorProps>) {
        super(props);
        this.state = { hasError: false };
    }

    componentDidCatch(error: Error, info: React.ErrorInfo): void {
        // Display fallback UI
        this.setState({ hasError: true, error: error, info: info });
        console.log('An exception occurred in a react component', error, info);
    }

    render(): React.ReactNode {
        if (this.state.hasError) {
            // You can render any custom fallback UI
            return (
                <div className="tr__panel">
                    <h1>{this.props.t('main:errors:ExceptionOnPage')}</h1>
                    <h3>{this.props.t('main:errors:ShowExceptionForCopy')}</h3>
                    <Collapsible trigger={this.props.t('main:errors:Exception')} open={false} transitionTime={100}>
                        <div className="tr__exception">
                            <span>{this.state.error && this.state.error.message}</span>
                            <span>{this.state.info && this.state.info.componentStack}</span>
                        </div>
                    </Collapsible>
                </div>
            );
        }
        return this.props.children;
    }
}

export default withTranslation('main')(ErrorBoundary);
