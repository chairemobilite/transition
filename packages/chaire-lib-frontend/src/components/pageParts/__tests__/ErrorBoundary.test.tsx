/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import * as React from 'react';
import { create } from 'react-test-renderer';
import { mount } from 'enzyme';
import ErrorBoundary from '../ErrorBoundary';

const testLabel = "Test";

test('Normal boundary', () => {
    const errorBoundary = create(<ErrorBoundary
        key={testLabel}>
            <input id="test"/>
        </ErrorBoundary>)
        .toJSON();
    expect(errorBoundary).toMatchSnapshot();
});

test('Boundary with error', () => {
    const errorMessage = 'This is an exception';
    // eslint-disable-next-line @typescript-eslint/ban-types
    const TestComponent: React.FunctionComponent<{}> = () => {
        throw new Error(errorMessage);
    };

    const errorBoundary = mount(<ErrorBoundary
        key={testLabel}>
            <TestComponent/>
        </ErrorBoundary>);
    // Make sure expected components are there
    const collapsible = errorBoundary.find('.Collapsible');
    expect(collapsible).toHaveLength(1);

    const trException = errorBoundary.find('.tr__exception');
    expect(trException).toHaveLength(1);

    // The trException's content should be the exception thrown
    const children = trException.children();
    expect(children.length).toEqual(2);
    expect(trException.childAt(0).text()).toEqual(errorMessage);
    expect(trException.childAt(1).text()).toContain("TestComponent");
});
