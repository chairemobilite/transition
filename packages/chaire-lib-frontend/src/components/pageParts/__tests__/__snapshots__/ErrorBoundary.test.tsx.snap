// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Normal boundary 1`] = `
<input
  id="test"
/>
`;
